### PR TITLE
Extend package to accept all deployment location parameters and other enhancements

### PR DESCRIPTION
**Pull request by eflamini**
- Added public network property into the assembly to be able to customize the name of floating IPs network.
- Added management of SSL client certificates and keys
- Added the validate_certs parameter commented out to be able to disable certificate validation
- Corrected scope in the get token API to be compatible with latest versions of identity API.
- Added adopt scenario example
- Restructured documentation

### CODE REVIEW

**Comment by eflamini** [comment]
@bfarrell-ibm: I've added you as a reviewer to review the changes in the README files.

File: apache-ansible-demo/docs/content.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
+The resource package contains resource definitions and scripts that allow to manage an Apache server on OpenStack through the Ansible Lifecycle Driver.
+
+The package contains the descriptor of an assembly ([apache-ansible-demo](./Descriptor/assembly.yml)) and of a resource ([hw-apache](./Contains/hw-apache-vnfc/Definitions/lm/resource.yaml)).
```
Comment (bfarrell-ibm) [inline apache-ansible-demo/docs/content.md]: The links to the resource and assembly descriptor are incorrect. You need to go up one level e.g. (../Descriptor/assembly.yml)

File: apache-ansible-demo/docs/content.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
+The resource package contains resource definitions and scripts that allow to manage an Apache server on OpenStack through the Ansible Lifecycle Driver.
+
+The package contains the descriptor of an assembly ([apache-ansible-demo](./Descriptor/assembly.yml)) and of a resource ([hw-apache](./Contains/hw-apache-vnfc/Definitions/lm/resource.yaml)).  
+The resource descriptor defines properties of the Apache server and of the underlying infrastructure. It also defines the characteristics of the lifecycle execution that is used to manage the resource.  
+The lifecycle definition indicates that Adopt, Create, Configure, Install, Start, Stop and Delete intents are defined for the resource and that are managed by the Ansible Lifecycle Driver.
```
Comment (bfarrell-ibm) [inline apache-ansible-demo/docs/content.md]: Is it more accurate to refer to these as Adopt, Create, Configure, Install, Start, Stop and Delete **transitions** ,rather than intents?

File: apache-ansible-demo/docs/content.md [hunk 3]
```diff
@@ -0,0 +1,13 @@
+The resource package contains resource definitions and scripts that allow to manage an Apache server on OpenStack through the Ansible Lifecycle Driver.
+
+The package contains the descriptor of an assembly ([apache-ansible-demo](./Descriptor/assembly.yml)) and of a resource ([hw-apache](./Contains/hw-apache-vnfc/Definitions/lm/resource.yaml)).  
+The resource descriptor defines properties of the Apache server and of the underlying infrastructure. It also defines the characteristics of the lifecycle execution that is used to manage the resource.  
+The lifecycle definition indicates that Adopt, Create, Configure, Install, Start, Stop and Delete intents are defined for the resource and that are managed by the Ansible Lifecycle Driver.  
+  
+The OpenStack infrastructure that hosts the Apache server is defined by the [HEAT template](./Contains/hw-apache-vnfc/Lifecycle/ansible/scripts/Openstack/heat.yaml).
```
Comment (bfarrell-ibm) [inline apache-ansible-demo/docs/content.md]: As above - update link

File: apache-ansible-demo/docs/content.md [hunk 4]
```diff
@@ -0,0 +1,13 @@
+The resource package contains resource definitions and scripts that allow to manage an Apache server on OpenStack through the Ansible Lifecycle Driver.
+
+The package contains the descriptor of an assembly ([apache-ansible-demo](./Descriptor/assembly.yml)) and of a resource ([hw-apache](./Contains/hw-apache-vnfc/Definitions/lm/resource.yaml)).  
+The resource descriptor defines properties of the Apache server and of the underlying infrastructure. It also defines the characteristics of the lifecycle execution that is used to manage the resource.  
+The lifecycle definition indicates that Adopt, Create, Configure, Install, Start, Stop and Delete intents are defined for the resource and that are managed by the Ansible Lifecycle Driver.  
+  
+The OpenStack infrastructure that hosts the Apache server is defined by the [HEAT template](./Contains/hw-apache-vnfc/Lifecycle/ansible/scripts/Openstack/heat.yaml).  
+  
+The lifecycle execution is implemented by [the ansible playbooks](./Contains/hw-apache-vnfc/Lifecycle/ansible/scripts). When an intent is issued against the resource, the Ansible Lifecycle Driver invokes the corresponding ansible playbook.
```
Comment (bfarrell-ibm) [inline apache-ansible-demo/docs/content.md]: As above - update link

File: apache-ansible-demo/docs/content.md [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+The ansible playbooks use artefacts in the package and properties passed by the Ansible Lifecycle Driver to execute the lifecycle transition.  
+For example, to create the OpenStack infrastructure, the Create playbook uses the HEAT template and passes Ansible Lifecycle Driver properties as the input parameters to the template.
+  
+A set of [bash scripts](./Contains/hw-apache-vnfc/Lifecycle/scripts) allow to invoke the ansible playbooks from command line. This can be useful to test alterations to the playbooks before installing the resource package. For details on how to run the bash scripts, refer to [the documentation](./scripts.md).
```
Comment (bfarrell-ibm) [inline apache-ansible-demo/docs/content.md]: As above - update link

File: apache-ansible-demo/docs/prerequisites.md [hunk 6]
```diff
@@ -0,0 +1,9 @@
+To run this resource package, you require the following prerequisites:
+
+1. An OpenStack environment configured with a user, an associated project, a deallocated floating IP, an Ubuntu 18.04 QCOW2 images and a *Key Pair* called *default* available on the target project.  
+2. For the Adopt lifecycle transition, an OpenStack stack in one of the following state: 'CREATE_COMPLETE','ADOPT_COMPLETE','RESUME_COMPLETE','CHECK_COMPLETE','UPDATE_COMPLETE'.  
+3. * A flavor of specification m1.small defined in the OpenStack project. The flavor defines the compute, memory, and storage capacity of a virtual server, also known as an instance.
+
+NOTE: Depending on your OpenStack HTTPS configuration, you may need to disable TLS certificates validation in the ansible playbooks. To do so, edit the files [Create.yaml](./Contains/hw-apache-vnfc/Lifecycle/ansible/scripts/Create.yaml), [Delete.yaml](./Contains/hw-apache-vnfc/Lifecycle/ansible/scripts/Delete.yaml) and [Adopt.yaml](./Contains/hw-apache-vnfc/Lifecycle/ansible/scripts/Adopt.yaml) and uncomment the *validate_certs* parameter.
```
Comment (bfarrell-ibm) [inline apache-ansible-demo/docs/prerequisites.md]: Update links to yaml files - should be folder one level above

**Review by bfarrell-ibm — CHANGES_REQUESTED** [review]
Links to some files not working - need to be updated. 
Also one comment to change 'intent' to 'transition' when describing the lifecycle section of the resource file.

**Review by owen-lynch-ibm — APPROVED** [review]
Changes for original commit in PR seem fine (Extend package to accept all dl parameters ) - I have not reviewed doc changes.

File: apache-ansible-demo/docs/prerequisites.md [hunk 7]
```diff
@@ -0,0 +1,9 @@
+To run this resource package, you require the following prerequisites:
+
+1. An OpenStack environment configured with a user, an associated project, a deallocated floating IP, an Ubuntu 18.04 QCOW2 images and a *Key Pair* called *default* available on the target project.  
+2. For the Adopt lifecycle transition, an OpenStack stack in one of the following state: 'CREATE_COMPLETE','ADOPT_COMPLETE','RESUME_COMPLETE','CHECK_COMPLETE','UPDATE_COMPLETE'.  
+3. * A flavor of specification m1.small defined in the OpenStack project. The flavor defines the compute, memory, and storage capacity of a virtual server, also known as an instance.
```
Comment (bfarrell-ibm) [inline apache-ansible-demo/docs/prerequisites.md]: Quite trivial - but there is a bullet point after the point 3 which is not necessary.

File: apache-ansible-demo/docs/scripts.md [hunk 8]
```diff
@@ -0,0 +1,13 @@
+The set of the scripts in this folder allows to invoke the resource package's playbooks from command line.
```
Comment (bfarrell-ibm) [inline apache-ansible-demo/docs/scripts.md]: Should there be a link here to ```../Contains/hw-apache-vnfc/Lifecycle/scripts``` folder? On current reading you would expect the scripts to be under the docs folder.

**Review by bfarrell-ibm — CHANGES_REQUESTED** [review]

File: apache-ansible-demo/docs/prerequisites.md [hunk 9]
```diff
@@ -0,0 +1,9 @@
+To run this resource package, you require the following prerequisites:
```
Comment (bfarrell-ibm) [inline apache-ansible-demo/docs/prerequisites.md]: Could this sentence be phrased better?
For example
To run the bash scripts and execute the example adopt procedure contained in this resource package, you require the following prerequisites:

**Review by bfarrell-ibm — CHANGES_REQUESTED** [review]

File: apache-ansible-demo/docs/adopt.md [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+### Procedure
+
+1. Download the apache-ansible-demo resource package:  
+```
+git clone https://github.com/IBM/tnco-package-examplesCopy code
+```
+
+2. Create an OpenStack stack that can be adopted by an assembly:  
+```
+cd tnco-package-examples/apache-ansible-demo/Contains/hw-apache-vnfc/Lifecycle/scripts/Copy code
```
Comment (bfarrell-ibm) [inline apache-ansible-demo/docs/adopt.md]: 'Copy code' text in the command lines should be removed.

**Review by bfarrell-ibm — CHANGES_REQUESTED** [review]

File: apache-ansible-demo/docs/adopt.md [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+   1. Set the TNCO_URL environment variable to point to the API URL:  
+   ```
+   export TNCO_URL=<ishtar_route>
+   ```  
+   2. Set the TNCO_TOKEN environment variable to point to the access token
+   ```
+   export TNCO_TOKEN=<access_token>
+   ```
+   
+   For more information about how to retrieve the ishtar_route and access_token parameters, see Invoking REST API.
```
Comment (bfarrell-ibm) [inline apache-ansible-demo/docs/adopt.md:82]: Should we add a link here to the KC section for Invoking the REST API
https://www-03preprod.ibm.com/support/knowledgecenter/SSDSDC_1.3/Installing/c_alm_invoking_rest_api.html

Comment (bfarrell-ibm) [inline apache-ansible-demo/docs/adopt.md]: Should we add a link here to the KC section for Invoking the REST API
https://www-03preprod.ibm.com/support/knowledgecenter/SSDSDC_1.3/Installing/c_alm_invoking_rest_api.html

**Review by bfarrell-ibm — CHANGES_REQUESTED** [review]

File: apache-ansible-demo/docs/adopt.md [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+```
+cd tnco-package-examples/apache-ansible-demo/Contains/hw-apache-vnfc/Lifecycle/scripts/Copy code
+```
+
+Edit the script file create.sh and customize the uncommented variables in the script:
+```
+OS_URL=<openstack-server-url>
+OS_AUTH_PROJECT_NAME=<openstack-project>
+OS_AUTH_USERNAME=<openstack-username>
+OS_AUTH_PASSWORD=<password>Copy code
```
Comment (bfarrell-ibm) [inline apache-ansible-demo/docs/adopt.md]: Remove 'Copy code' text

File: apache-ansible-demo/docs/adopt.md [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+cd ../../
+zip -r  apache-ansible-demo .
+curl -sk -X POST "https://$TNCO_URL/api/resource-manager/resource-packages" -H 'Authorization: Bearer '"${TNCO_TOKEN}"'' -H 'Content-Type: multipart/form-data' -F "file=@./apache-ansible-demo.zip"
+```  
+8. Upload the resource and assembly descriptors to the IBM Telco Network Cloud - Orchestration instance:
+```
+cd Definitions/lm
+curl -sk -H 'Accept: */*' -X POST -H 'Content-Type: application/yaml' -H 'Authorization: Bearer '"${TNCO_TOKEN}"'' "https://$TNCO_URL/api/catalog/descriptors" --data-binary @./resource.yaml
+cd ../../../../Descriptor
+curl -sk -H 'Accept: */*' -X POST -H 'Content-Type: application/yaml' -H 'Authorization: Bearer '"${TNCO_TOKEN}"'' "https://$TNCO_URL/api/catalog/descriptors" --data-binary @./assembly.ymlCopy code
```
Comment (bfarrell-ibm) [inline apache-ansible-demo/docs/adopt.md]: Remove 'Copy code' text

**Review by bfarrell-ibm — CHANGES_REQUESTED** [review]

File: apache-ansible-demo/docs/prerequisites.md (hunk 9 above)
Comment (eflamini) [reply apache-ansible-demo/docs/prerequisites.md]: I'm going to rephrase this, but based on the following considerations:
```
Those are the prerequisites to run the resource package in ALM. Not only for the adopt example but in general if you want to use the resource package to create, delete start and stop the network service through ALM.

The prerequisites in the adopt example have a link to them as they apply to it too.

About the bash scripts, the scripts.md should articulate the requirements for them separately.
```

File: apache-ansible-demo/docs/adopt.md (hunk 11 above)
Comment (eflamini) [reply apache-ansible-demo/docs/adopt.md]: I'm not a fan of links. Especially when they are tight to a version of the product or then they can change. I'm going to rephrase it to give more indication of where to find that documentation and I'm going to put a generic link to the TNCO knowledge center (so that it is valid also for future versions).

File: apache-ansible-demo/docs/content.md [hunk 14]
```diff
@@ -0,0 +1,13 @@
+The resource package contains resource definitions and scripts that allow to manage an Apache server on OpenStack through the Ansible Lifecycle Driver.
+
+The package contains the descriptor of an assembly ([apache-ansible-demo](../Descriptor/assembly.yml)) and of a resource ([hw-apache](../Contains/hw-apache-vnfc/Definitions/lm/resource.yaml)).  
+The resource descriptor defines properties of the Apache server and of the underlying infrastructure. It also defines the characteristics of the lifecycle execution that is used to manage the resource.  
+The lifecycle definition indicates that Adopt, Create, Configure, Install, Start, Stop and Delete lifecycle transitions are defined for the resource and that are managed by the Ansible Lifecycle Driver.  
+  
+The OpenStack infrastructure that hosts the Apache server is defined by the [HEAT template](../Contains/hw-apache-vnfc/Lifecycle/ansible/scripts/Openstack/heat.yaml).  
+  
+The lifecycle execution is implemented by [the ansible playbooks](../Contains/hw-apache-vnfc/Lifecycle/ansible/scripts). When an intent is issued against the resource, the Ansible Lifecycle Driver invokes the corresponding ansible playbook.
```
Comment (sglover) [inline apache-ansible-demo/docs/content.md]: When an intent is issued  -> When a lifecycle transition is executed

**Review by sglover — APPROVED** [review]
Nothing to add that Brendan hasn't already mentioned

**Comment by owen-lynch-ibm** [comment]
@bfarrell-ibm have all changes requested been addressed here ?

**Comment by bfarrell-ibm** [comment]
Yes - i think all issues i raised have been resolved.

**Review by bfarrell-ibm — APPROVED** [review]
